File: SortingSearching_Java_RiyaBansal/L7/Ternary_Search.java
package Sorting_Searching_7;

public class Ternary_Search {

	public static void main(String[] args) {
		

		int[] arr = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
		
		System.out.println(ternarySearch(arr, 9));
		

	}

	private static int  ternarySearch(int[] arr, int key) {
		
		int l = 0;
		int r = arr.length - 1;
		while(r >= l) {
			
			int mid1 = l + (r - l) / 3;
			int mid2 = r - (r - l) / 3;
			
			if(arr[mid1] == key) {
				return mid1;
			}
			
			if(arr[mid2] == key) {
				return mid2;
			}
			
			if(key < arr[mid1]) {
				r = mid1 - 1;
			}else if(key > arr[mid2]) {
				l = mid2 + 1;
			}else {
				
				l = mid1 + 1;
				r = mid2 - 1;
			}
			
		}
		
		return -1;
		
	}

}

File: sorting_and_searching_in_python/Lecture15/ternary_search.py
def ternary_search(arr, el):
	lo, hi = 0, len(arr) - 1
	while lo <= hi:
		m1 = lo + (hi - lo) // 3
		m2 = hi - (hi - lo) // 3
		if arr[m1] == el:
			return m1
		if arr[m2] == el:
			return m2
		if arr[m1] > el:
			hi = m1 - 1
		elif arr[m2] < el:
			lo = m2 + 1
		else:
			lo = m1 + 1
			hi = m2 - 1

	return -1


